### ISSUE TITLE
feat(inference): Flash Attention v2 with variable-length sequence optimization

### ISSUE THREAD

**Issue by mrveiss** [mrveiss/AutoBot-AI#1955]
## Summary

Integrate Flash Attention v2 with variable-length sequence support for batch inference, using unpadding to avoid wasting GPU compute on padding tokens.

## Background

When processing multiple requests simultaneously, sequences have different lengths and must be padded to the longest. Standard attention wastes compute on padding tokens. airllm's approach:

1. **No padding in batch** → `flash_attn_kvpacked_func()` directly (fastest path)
2. **Padding present** → `unpad_input()` removes padding → `flash_attn_varlen_kvpacked_func()` processes only real tokens → `pad_input()` restores batch shape

Additional optimizations from airllm:
- **Fused RoPE kernel**: `apply_rotary_emb_func` from flash_attn (avoids separate kernel launch)
- **JIT-compiled GQA expansion**: `@torch.jit.script` on `repeat_kv()` for Grouped Query Attention
- **Growing KV cache**: extend by 256 positions at a time (`torch.cat([past_kv, torch.empty(bsz, 256, ...)])`) instead of reallocating per token

```python
from flash_attn import flash_attn_kvpacked_func, flash_attn_varlen_kvpacked_func
from flash_attn.bert_padding import unpad_input, pad_input

if key_padding_mask is None:
    # No padding — fastest path
    output = flash_attn_kvpacked_func(q, kv, dropout_p, softmax_scale, causal=True)
else:
    # Variable-length — unpad, compute, repad
    nheads = q.shape[2]
    q_unpad, indices, cu_seqlens, max_seqlen = unpad_input(q, key_padding_mask)
    kv_unpad = unpad_input(kv, key_padding_mask)
    output_unpad = flash_attn_varlen_kvpacked_func(
        q_unpad, kv_unpad, cu_seqlens, cu_seqlens, max_seqlen, max_seqlen,
        dropout_p, softmax_scale, causal=True)
    output = pad_input(output_unpad, indices, batch_size, max_seqlen)
```

## Key Files

- `autobot-backend/llm_interface_pkg/optimization/flash_attention.py` — new
- `autobot-backend/llm_interface_pkg/providers/` — integrate into attention computation
- Related: attention fallback chain issue (Flash Attn is Tier 1)

## Acceptance Criteria

- [ ] Flash Attention v2 integrated with two paths (padded/unpadded)
- [ ] Variable-length sequence support via `flash_attn_varlen_kvpacked_func`
- [ ] Fused RoPE kernel from flash_attn package
- [ ] Growing KV cache (256-position chunks) to reduce allocation overhead
- [ ] JIT-compiled `repeat_kv` for GQA models
- [ ] Works on RTX 4070 with CUDA 12
- [ ] Graceful fallback when flash_attn not installed (→ SDPA or vanilla)
- [ ] Benchmark: batch of 4 requests with varying lengths — compute savings vs padded attention

## Related Issues

- Attention fallback chain issue (Flash Attn is the top tier)
- #1946 (layer-by-layer inference — Flash Attn applies per-layer)
